.github/workflows/kind-e2e.yaml

```diff
@@ -203,6 +208,59 @@ jobs:
         echo "GATEWAY_OVERRIDE=$GATEWAY_OVERRIDE" >> $GITHUB_ENV
         echo "GATEWAY_NAMESPACE_OVERRIDE=$GATEWAY_NAMESPACE_OVERRIDE" >> $GITHUB_ENV
 
+    - name: Apply NetworkPolicy
+      working-directory: ./src/knative.dev/serving
+      run: |
+        set -x
+
+        cat >> netpolicy.yaml <<EOF
+        apiVersion: networking.k8s.io/v1
+        kind: NetworkPolicy
+        metadata:
+          name: default-deny
+          namespace: $SYSTEM_NAMESPACE
+        spec:
+          podSelector:
+            matchExpressions:
+            - key: app
+              operator: In
+              values: [autoscaler, autoscaler-hpa]
+          policyTypes:
+          - Ingress
+          ingress:
+          - from:
+            - namespaceSelector:
+                matchLabels:
+                  # kubernetes.io/metadata.name: $SYSTEM_NAMESPACE # this only works on k8s >= 1.21.1
+                  projectcalico.org/name: $SYSTEM_NAMESPACE # calico specific
+        EOF
+        kubectl apply -f netpolicy.yaml
+
+    - name: Check that autoscaler pods are unreachable
+      run: |
+        set -x
+
+        AUTOSCALER_IP=$(kubectl -n $SYSTEM_NAMESPACE get po -l 'app=autoscaler' -o jsonpath='{.items[*].status.podIPs[*].ip}')
+
+        # ping from inside the installation namespace should succeed
+        kubectl run pinger --rm -i --image=busybox --namespace=$SYSTEM_NAMESPACE --restart=Never -- ping -c 8 $AUTOSCALER_IP
+
+        PING_FAILED=false
+
+        function catch_ping_err {
+          if $PING_FAILED; then
+            echo "pinging autoscaler failed as expected"
+          else
+            echo "pingign autoscaler should have failed but succeded"
+            exit 1
+          fi
+        }
+
+        trap catch_ping_err ERR
```

Comment:
this is probably just me not understanding bash well enough, but I'm not sure why we need a trap here rather than just something like:

~~~~bash
if kubectl run pinger --rm -i --image=busybox --restart=Never -- ping -c 8 $AUTOSCALER_IP; then
  echo "Expected ping to autoscaler to fail, but it succeeded"
fi
~~~~

Reply:
I think if it fails without the trap the script just fails and that's what made  me have to do it this way but let me see if I that works